src/rust/sqs-executor/src/retriever/event_retriever.rs

```diff
@@ -0,0 +1,231 @@
+{
+    type Message = SqsMessage;
+    type Error = S3PayloadRetrieverError<DecoderErrorT>;
+
+    #[tracing::instrument(skip(self, msg))]
+    async fn retrieve_event(&mut self, msg: &Self::Message) -> Result<Option<E>, Self::Error> {
+        let body = msg.body.as_ref().unwrap();
+        println!("Got body from message: {}", body);
+        debug!("Got body from message: {}", body);
+        let event: serde_json::Value = serde_json::from_str(body)?;
+
+        if let Some(Some(event_str)) = event.get("Event").map(serde_json::Value::as_str) {
+            if event_str == "s3:TestEvent" {
+                return Ok(None);
+            }
+        }
+        let record = &event["Records"][0]["s3"];
+
+        let bucket = record["bucket"]["name"].as_str().expect("bucket name");
+        let key = record["object"]["key"].as_str().expect("object key");
+        debug!("Retrieving S3 payload from: {} / {}", bucket, key);
+        let region = &event["Records"][0]["awsRegion"].as_str().expect("region");
+        let s3 = self.get_client(region.to_string());
+        let s3_data = s3.get_object(GetObjectRequest {
+            bucket: bucket.to_string(),
+            key: key.to_string(),
+            ..Default::default()
+        });
+
+        let s3_data = tokio::time::timeout(Duration::from_secs(3), s3_data);
+        let s3_data = s3_data
+            .timed()
+            .map(|(s3_data, ms)| {
+                self.metric_reporter
+                    .histogram(
+                        "s3_consumer.get_object.ms",
+                        ms as f64,
+                        &[tag("success", s3_data.is_ok())],
+                    )
+                    .unwrap_or_else(|e| {
+                        error!("failed to report s3_consumer.get_object.ms: {:?}", e)
+                    });
+                s3_data
+            })
+            .await??;
+
+        let object_size = record["object"]["size"].as_u64().unwrap_or_default();
+        let prealloc = if object_size < 1024 {
+            1024
+        } else {
+            object_size as usize
+        };
+
+        info!("Retrieved s3 payload with size : {:?}", prealloc);
+
+        let mut body = Vec::with_capacity(prealloc);
+
+        s3_data
+            .body
+            .expect("Missing S3 body")
+            .into_async_read()
+            .read_to_end(&mut body)
+            .timed()
+            .map(|(res, ms)| {
+                self.metric_reporter
+                    .histogram(
+                        "s3_consumer.read_to_end.ms",
+                        ms as f64,
+                        &[tag("success", res.is_ok())],
+                    )
+                    .unwrap_or_else(|e| {
+                        error!("failed to report s3_consumer.read_to_end.ms: {:?}", e)
+                    });
+                res
+            })
+            .await?;
+
+        self.metric_reporter
+            .gauge("s3_retriever.bytes", body.len() as f64, &[])
```

Comment:
previous histograms use verbiage 's3_consumer' not s3_retriever; just make em match (unless this is intentional)

Reply:
also line below

Reply:
also line 222